scoap3/modules/tools/tasks.py

```diff
@@ -0,0 +1,109 @@
+    :param tool_name: name of the tool, which will be used in the subject of the email.
+    """
+    timestamp = datetime.now().strftime('%Y-%m-%dT%H:%M:%S')
+    filename = 'scoap3_export_%s_%s.csv' % (tool_name, timestamp)
+    attachment = Attachment(filename=filename, content_type=content_type, data=result_data)
+
+    msg = TemplatedMessage(
+        template_html='scoap3_tools/email/result.html',
+        subject='SCOAP3 - Export %s result' % tool_name,
+        sender=current_app.config.get('MAIL_DEFAULT_SENDER'),
+        recipients=recipients,
+        attachments=[attachment],
+    )
+    current_app.extensions['mail'].send(msg)
+
+
+def send_failed_email(recipients, tool_name, task_id=None):
+    """
+    Notifies the user about a failed generation.
+    :param recipients: recipients who will receive the email.
+    :param tool_name: name of the tool, which will be used in the subject of the email.
+    :param task_id: celery task id, if available.
+    """
+    msg = TemplatedMessage(
+        template_html='scoap3_tools/email/failed.html',
+        subject='SCOAP3 - Export %s result error' % tool_name,
+        sender=current_app.config.get('MAIL_DEFAULT_SENDER'),
+        recipients=recipients,
+        ctx={'task_id': task_id}
+    )
+    current_app.extensions['mail'].send(msg)
+
+
+@shared_task(bind=True)
+def run_tool(self, result_email, tool_name, serializer_function=to_csv, **kwargs):
+    """
+    Wrapper for generating result for a tool.
+
+    It generates the result using the tool_function parameter's return value, then sends it via email.
+    :param self: bound task type instance.
+    :param result_email: email address to send the results to.
+    :param tool_name: name of the tool, which is used to determine the generator function.
+    :param serializer_function: serializer function
+    :param kwargs: additional kwargs passed to the tool_function
+    """
+    try:
+        logger.info('Running tool. tool_name=%s result_email=%s' % (tool_name, result_email))
+
+        tool_function = current_app.config.get('TOOL_FUNCTIONS', {}).get(tool_name)
+        if tool_function is None:
+            logger.error('Invalid tool name: %s' % tool_name)
+            send_failed_email([result_email], tool_name)
+            return
+
+        result = tool_function(**kwargs)
+        logger.info('Result calculated. result_data_count=%d tool_name=%s result_email=%s' % (
+            len(result['data']), tool_name, result_email))
+
+        content_type, serialized_result = serializer_function(result)
+        logger.info('Result serialized, sending email... result_data_count=%d tool_name=%s result_email=%s' % (
+            len(result['data']), tool_name, result_email))
+
+        send_result(serialized_result, content_type, [result_email], tool_name)
+
+        logger.info('Result successfully sent. tool_name=%s result_email=%s' % (tool_name, result_email))
+    except Exception:
```

Comment:
I would put `result = tool_function(**kwargs)` and ` send_result` in twp different try/except blocks.
If the task succeeds and the email fail, we might want to communicate it (i.e. if the attachment is too big, then just provide a hint for retrieving the csv manually from somewhere)

Reply:
Of course, internally :point_up: we can also think about alternative ideas maybe

Reply:
:+1: